addons/purchase/models/account_invoice.py

```diff
@@ -81,23 +81,27 @@ def _onchange_purchase_auto_complete(self):
     @api.onchange('partner_id', 'company_id')
     def _onchange_partner_id(self):
         res = super(AccountMove, self)._onchange_partner_id()
-        if self.partner_id and\
-                self.move_type in ['in_invoice', 'in_refund'] and\
-                self.currency_id != self.partner_id.property_purchase_currency_id and\
-                self.partner_id.property_purchase_currency_id.id:
+
+        currency_id = (
+                self.partner_id.property_purchase_currency_id
+                or self.env.context.get("default_currency_id")
+                or self.env.company.currency_id
+        )
+
+        if self.partner_id and self.move_type in ['in_invoice', 'in_refund'] and self.currency_id != currency_id:
             if not self.env.context.get('default_journal_id'):
                 journal_domain = [
                     ('type', '=', 'purchase'),
                     ('company_id', '=', self.company_id.id),
-                    ('currency_id', '=', self.partner_id.property_purchase_currency_id.id),
+                    ('currency_id', '=', currency_id.id),
                 ]
                 default_journal_id = self.env['account.journal'].search(journal_domain, limit=1)
                 if default_journal_id:
                     self.journal_id = default_journal_id
-            if self.env.context.get('default_currency_id'):
-                self.currency_id = self.env.context['default_currency_id']
-            if self.partner_id.property_purchase_currency_id:
-                self.currency_id = self.partner_id.property_purchase_currency_id
+
+            self.currency_id = currency_id
+            self.line_ids.currency_id = currency_id
```

Comment:
why not calling self._onchange_currency() instead?

Reply:
We could do 
```suggestion
            self._onchange_currency()
```
It can indeed take into account other logics